core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1019,16 +1019,31 @@ public Builder setBranchSnapshot(long snapshotId, String branch) {
       return this;
     }
 
-    public Builder removeBranch(String branch) {
-      if (SnapshotRef.MAIN_BRANCH.equals(branch)) {
+    public Builder setRef(String name, SnapshotRef ref) {
+      SnapshotRef existingRef = refs.get(name);
+      if (existingRef != null && existingRef.equals(ref)) {
+        return this;
+      }
+
+      long snapshotId = ref.snapshotId();
+      Snapshot snapshot = snapshotsById.get(snapshotId);
+      ValidationException.check(snapshot != null, "Cannot set %s to unknown snapshot: %s", name, snapshotId);
+      refs.put(name, ref);
```

Comment:
This will discard existing ref settings, like `minSnapshotsToKeep`.

I think that carrying existing ref settings through should be thoroughly tested.